velox/type/StringToFloatParser.cpp

```diff
@@ -0,0 +1,202 @@
+                  return true;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+static inline bool isNanString(const char* str, size_t length) {
+  if (length != 3) {
+    return false;
+  }
+
+  if (isEither(str, 0, 'n', 'N')) {
+    if (isEither(str, 1, 'a', 'A')) {
+      if (isEither(str, 2, 'n', 'N')) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+template <typename T>
+Status StringToFloatParser::parse(const std::string_view& str, T& out) {
+  auto length = str.length();
+  if (UNLIKELY(length <= 0)) {
+    return Status::Invalid("empty string");
+  }
+  int i = 0;
+  // skip any spaces at the start
+  for (; i < length; ++i) {
+    if (!characterIsSpace(str[i])) {
+      break;
+    }
+  }
+
+  // skip any spaces at the end
+  int j = length - 1;
+  for (; j >= i; j--) {
+    if (!characterIsSpace(str[j])) {
+      break;
+    }
+  }
+
+  if (UNLIKELY(i > j)) {
+    return Status::Invalid("i > j");
+  }
+
+  bool negative = false;
+  // skip leading +/-
+  switch (str[i]) {
+    case '-':
+      negative = true;
+      i++;
+      break;
+    case '+':
+      i++;
+  }
+
+  if (UNLIKELY(i > j)) {
+    return Status::Invalid("i > j");
+  }
+
+  // check INF/INFINITY/NAN, ignoring case
+  for (int k = i; k <= j; ++k) {
+    if (str[k] == 'i' || str[k] == 'I') {
+      if (isInfinityString(str.data() + k, length - k)) {
+        out = negative ? -INFINITY : INFINITY;
+        return Status::OK();
+      } else {
+        // Starts with 'i', but isn't inf.
+        return Status::Invalid("Invalid input string: {}", str);
+      }
+    } else if (str[k] == 'n' || str[k] == 'N') {
+      if (isNanString(str.data() + k, length - k)) {
+        out = negative ? -NAN : NAN;
+        return Status::OK();
+      } else {
+        // Starts with 'n', but isn't NaN.
+        return Status::Invalid("Invalid input string: {}", str);
+      }
+    }
+  }
+
+  // check invalid char
+  bool exponential = false;
+  bool point = false;
+  bool invalid = false;
+  for (int k = i; k <= j; ++k) {
+    if (str[k] >= '0' && str[k] <= '9') {
+      continue;
+    } else if (str[k] == '.') {
+      if (LIKELY(!point)) {
+        point = true;
+      } else {
+        invalid = true;
+        break;
+      }
+    } else if (str[k] == 'e' || str[k] == 'E') {
+      if (LIKELY(!exponential)) {
+        exponential = true;
+      } else {
+        invalid = true;
+        break;
+      }
+    } else if (str[k] == '-' || str[k] == '+') {
+      if (LIKELY(k > i && (str[k - 1] == 'e' || str[k - 1] == 'E'))) {
+        continue;
+      } else {
+        invalid = true;
+        break;
+      }
+    } else {
+      invalid = true;
+      break;
+    }
+  }
+
+  if (invalid) {
+    return Status::Invalid(
+        "Non-whitespace character found after end of conversion");
+  }
+
+  double val;
+  auto res = fast_float::from_chars(str.data() + i, str.data() + j + 1, val);
```

Comment:
Is the logic up to this line a copy from folly::to? If so, would it make sense to change folly::to? Does it use std::from_chars now?

Reply:
No, In fact, this method heavily references the implementation of [starrocks](https://github.com/StarRocks/starrocks/commit/b3b4263d28f88cd7dc251679aba228ad1ea65d0f). `folly::to` does not use `std::from_chars`.